### PR TITLE
feat: remove `deploy` instruction from venom

### PR DESCRIPTION
**Pull request by harkal**
### What I did

Venom IR for an easier transition employed a ```deploy``` instruction like the old IR. This was creating friction in the new system as it was implicitly dictating basic block flows, that had to be handled with special cases code. 

This PR removes the need for the ```deploy``` instruction. The original IR is split into deploy and runtime code as soon as possible, and the two parts are handed to the new backend, which now handles any number of contexts that it links when emitting assembly.

### Commit message

Removes the need for the ```deploy``` instruction
* Venom IR now does not support the ```deploy``` instruction
* Makes the venom backend handle multiple contexts (functions to be linked together)
* Adds multiple entry points to contexts

### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

File: vyper/venom/__init__.py [hunk 1]
```diff
@@ -19,17 +19,14 @@
 
 
 def generate_assembly_experimental(
-    ctx: IRFunction, optimize: Optional[OptimizationLevel] = None
+    ctxs: tuple[IRFunction, IRFunction], optimize: Optional[OptimizationLevel] = None
```
Comment (charles-cooper) [inline vyper/venom/__init__.py:22]: this should either be a list of IRFunctions or specify by names -- runtime_ir and deploy_ir.

File: vyper/venom/venom_to_assembly.py [hunk 2]
```diff
@@ -75,13 +77,13 @@
 # with the assembler. My suggestion is to let this be for now, and we can
 # refactor it later when we are finished phasing out the old IR.
 class VenomCompiler:
-    ctx: IRFunction
+    ctxs: list[IRFunction]
```
Comment (charles-cooper) [inline vyper/venom/venom_to_assembly.py:80]: should these still be named `ctxs`? maybe `functions`?

**Review by charles-cooper — COMMENTED** [review]
seems fine, but not a fan of the new API for `generate_assembly_experimental`. (ditto for `VenomCompiler` -- i think it should not allow `None` functions as part of its input.

**Comment by harkal** [comment]
> seems fine, but not a fan of the new API for `generate_assembly_experimental`. (ditto for `VenomCompiler` -- i think it should not allow `None` functions as part of its input.

generate_assembly_experimental() is just glue code for the old-new code interface. It's basic purpose is to "untangle" the combination of ir_runtime and ir_nodes properties that depending on the run parameters can be multiple things (one of them being the (None, IRnode) tuple in the case of doing runtime only code. 

VenomCompiler() takes just a list of IRFunctions to compiler, which is clean. With the last commit I removed the None check inside VenomCompiler for cleaning up the list of Nones in the glue code which is indeed nicer.

**Pull request merged**